### PR TITLE
Add database migration for the user_invites table

### PR DESCRIPTION
**Pull request by rdimitrov**
# Summary

The following adds a user_invites table.

The role column is a string where the expected values are some of 'admin', 'editor', 'viewer', 'policy_writer', 'permissions_manager' (at least for now).

Fixes: https://github.com/stacklok/minder/issues/3533

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [x] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [ ] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [ ] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

File: database/migrations/000061_user_invites.up.sql [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+CREATE TYPE user_role AS ENUM ('admin', 'editor', 'viewer', 'policy_writer', 'permissions_manager');
+
+CREATE TABLE IF NOT EXISTS user_invites(
+    id           UUID NOT NULL DEFAULT gen_random_uuid() PRIMARY KEY,
+    email        TEXT NOT NULL,
+    status       invite_status NOT NULL DEFAULT 'pending',
+    role         user_role NOT NULL DEFAULT 'viewer',
+    project      UUID NOT NULL REFERENCES projects(id) ON DELETE CASCADE,
+    invitee      INTEGER NULL REFERENCES users(id) ON DELETE CASCADE,
+    invited_by   INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
```
Comment (rdimitrov) [inline database/migrations/000061_user_invites.up.sql:29]: Noticed that this is SERIAL (not UUID) thus why the INTEGER type.

Comment (dmjb) [reply database/migrations/000061_user_invites.up.sql:29]: Might it make sense to add a UUID ID column to the users table, populate with autogenerated values, and then replace use of the integer ID with a UUID one?

Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql:29]: It's out of the scope of this effort but it's a good idea, i.e. I feel much better seeing uuids as IDs then 7 for example 😃

File: database/migrations/000061_user_invites.up.sql [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    project      UUID NOT NULL REFERENCES projects(id) ON DELETE CASCADE,
+    invitee      INTEGER NULL REFERENCES users(id) ON DELETE CASCADE,
+    sponsor      INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+    code         TEXT NOT NULL,
+    created_at   TIMESTAMP NOT NULL DEFAULT NOW(),
+    updated_at   TIMESTAMP NOT NULL DEFAULT NOW(),
+    -- ensure there's one invite for this email per project
+    UNIQUE (email, project),
+    -- ensure code/nonce is unique
+    UNIQUE (code)
```
Comment (evankanderson) [inline database/migrations/000061_user_invites.up.sql:36]: Can we use `code` as the primary key?

File: database/migrations/000061_user_invites.up.sql [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+-- describes the invite status
+CREATE TYPE invite_status AS ENUM ('pending', 'accepted', 'declined', 'expired', 'revoked');
+-- reflects the roles we have in OpenFGA
+CREATE TYPE user_role AS ENUM ('admin', 'editor', 'viewer', 'policy_writer', 'permissions_manager');
+
+CREATE TABLE IF NOT EXISTS user_invites(
+    id           UUID NOT NULL DEFAULT gen_random_uuid() PRIMARY KEY,
+    email        TEXT NOT NULL,
+    status       invite_status NOT NULL DEFAULT 'pending',
```
Comment (evankanderson) [inline database/migrations/000061_user_invites.up.sql:25]: What are other values of `status` would we use?  (Do we need this column?)

**Review by evankanderson — COMMENTED** [review]
A few comments to hopefully reduce the amount of work.  😁

File: database/migrations/000061_user_invites.up.sql (hunk 3 above)
Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql:25]: I was thinking of the use case where you list the invitations and see some that are pending, some expired, but it makes more sense to make that as separate audit log. Thanks!

File: database/migrations/000061_user_invites.up.sql [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    role         user_role NOT NULL DEFAULT 'viewer',
+    project      UUID NOT NULL REFERENCES projects(id) ON DELETE CASCADE,
+    invitee      INTEGER NULL REFERENCES users(id) ON DELETE CASCADE,
+    sponsor      INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+    code         TEXT NOT NULL,
+    created_at   TIMESTAMP NOT NULL DEFAULT NOW(),
+    updated_at   TIMESTAMP NOT NULL DEFAULT NOW(),
+    -- ensure there's one invite for this email per project
+    UNIQUE (email, project),
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql:34]: I had the assumption that a user can have only one role per project or this was incorrect?

**Comment by rdimitrov** [comment]
@evankanderson - Thanks for the tips! 🙏 I like how much cleaner it got after applying your feedback 👍

File: database/migrations/000061_user_invites.up.sql [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+BEGIN;
+
+-- describes the invite status
+CREATE TYPE invite_status AS ENUM ('pending', 'accepted', 'declined', 'expired', 'revoked');
+-- reflects the roles we have in OpenFGA
+CREATE TYPE user_role AS ENUM ('admin', 'editor', 'viewer', 'policy_writer', 'permissions_manager');
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql]: Addressed.

File: database/migrations/000061_user_invites.up.sql [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+CREATE INDEX idx_user_invites_email ON user_invites(email);
+
+-- create an index on the project column
+CREATE INDEX idx_user_invites_project ON user_invites(project);
+
+-- create an index on the invitee column
+CREATE INDEX idx_user_invites_invitee ON user_invites(invitee);
+
+-- create an index on the invited_by column
+CREATE INDEX idx_user_invites_sponsor ON user_invites(sponsor);
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql]: I was thinking of helping the use case for "give me all invitations I've received from X".

File: database/migrations/000061_user_invites.up.sql [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+-- reflects the roles we have in OpenFGA
+CREATE TYPE user_role AS ENUM ('admin', 'editor', 'viewer', 'policy_writer', 'permissions_manager');
+
+CREATE TABLE IF NOT EXISTS user_invites(
+    id           UUID NOT NULL DEFAULT gen_random_uuid() PRIMARY KEY,
+    email        TEXT NOT NULL,
+    status       invite_status NOT NULL DEFAULT 'pending',
+    role         user_role NOT NULL DEFAULT 'viewer',
+    project      UUID NOT NULL REFERENCES projects(id) ON DELETE CASCADE,
+    invitee      INTEGER NULL REFERENCES users(id) ON DELETE CASCADE,
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql]: Addressed

File: database/migrations/000061_user_invites.up.sql (hunk 2 above)
Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql]: Addressed

File: database/migrations/000061_user_invites.up.sql [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    created_at   TIMESTAMP NOT NULL DEFAULT NOW(),
+    updated_at   TIMESTAMP NOT NULL DEFAULT NOW(),
+    -- ensure there's one invite for this email per project
+    UNIQUE (email, project),
+    -- ensure code/nonce is unique
+    UNIQUE (code)
+);
+
+-- create an index on the email column
+CREATE INDEX idx_user_invites_email ON user_invites(email);
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply database/migrations/000061_user_invites.up.sql]: Addressed

File: database/migrations/000062_user_invites.up.sql [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+BEGIN;
+
+CREATE TABLE IF NOT EXISTS user_invites (
+    code          TEXT NOT NULL PRIMARY KEY,
+    email         TEXT NOT NULL,
```
Comment (evankanderson) [inline database/migrations/000062_user_invites.up.sql:19]: Removing the index on `(email, project)` means that we no longer have an index on email.  I'd be tempted to re-add this.

Comment (rdimitrov) [reply database/migrations/000062_user_invites.up.sql:19]: Ugh, you're right! I must have deleted it along the other stuff 🤦‍♂️

Comment (evankanderson) [reply database/migrations/000062_user_invites.up.sql]: I pointed out that a `UNIQUE(email, project)` index would prevent inviting someone to multiple roles in a project, so it was sort of my fault.

**Review by evankanderson — APPROVED** [review]

**Pull request merged**